vibrato/src/trainer.rs

```diff
@@ -1,4 +1,473 @@
+        let mut unk_handler_wtr = BufWriter::new(unk_handler_wtr);
+        let mut connector_wtr = BufWriter::new(connector_wtr);
+
+        let mut output = [0; 4096];
+
+        // scales weights to represent them in i16.
+        let mut weight_abs_max = 0f64;
+        for feature_set in &merged_model.feature_sets {
+            weight_abs_max = weight_abs_max.max(feature_set.weight.abs());
+        }
+        for hm in &merged_model.matrix {
+            for &w in hm.values() {
+                weight_abs_max = weight_abs_max.max(w);
+            }
+        }
+        let weight_scale_factor = f64::from(i16::MAX) / weight_abs_max;
+
+        for i in 0..self.surfaces.len() {
+            let mut writer = csv_core::Writer::new();
+            let mut surface = self.surfaces[i].as_bytes();
+            let feature_set = merged_model.feature_sets[i];
+            let word_idx = WordIdx::new(LexType::System, u32::try_from(i).unwrap());
+            let feature = self.dict.system_lexicon().word_feature(word_idx);
+
+            // writes surface
+            loop {
+                let (result, nin, nout) = writer.field(surface, &mut output);
+                lexicon_wtr.write_all(&output[..nout])?;
+                if result == csv_core::WriteResult::InputEmpty {
+                    break;
+                }
+                surface = &surface[nin..];
+            }
+            let (result, nout) = writer.finish(&mut output);
+            assert_eq!(result, csv_core::WriteResult::InputEmpty);
+            lexicon_wtr.write_all(&output[..nout])?;
+
+            // writes others
+            writeln!(
+                &mut lexicon_wtr,
+                ",{},{},{},{}",
+                feature_set.left_id,
+                feature_set.right_id,
+                (-feature_set.weight * weight_scale_factor) as i16,
+                feature,
+            )?;
+        }
+
+        for i in 0..self.dict.unk_handler().len() {
+            let word_idx = WordIdx::new(LexType::Unknown, u32::try_from(i).unwrap());
+            let cate_id = self.dict.unk_handler().word_cate_id(word_idx);
+            let feature = self.dict.unk_handler().word_feature(word_idx);
+            let cate_string = self
+                .dict
+                .char_prop()
+                .cate_string(u32::from(cate_id))
+                .unwrap();
+            let feature_set = merged_model.feature_sets[self.surfaces.len() + i];
+            writeln!(
+                &mut unk_handler_wtr,
+                "{},{},{},{},{}",
+                cate_string,
+                feature_set.left_id,
+                feature_set.right_id,
+                (-feature_set.weight * weight_scale_factor) as i16,
+                feature,
+            )?;
+        }
+
+        writeln!(
+            &mut connector_wtr,
+            "{} {}",
+            merged_model.right_conn_to_left_feats.len(),
+            merged_model.left_conn_to_right_feats.len(),
+        )?;
```

Comment:
This is just a confirmation.

The definition of the `matrix.def` is counterintuitive: The first column corresponds to **right** ids, and the second column corresponds to **left** ids. No problem in your code?

Reply:
No problem.
I changed the variable names to make this easy to understand: `i` -> `right_conn_id`, `j` -> `left_conn_id`.